### ISSUE TITLE
Virtual root support could be a vector for cache poisoning attacks

### ISSUE THREAD

**Issue by nickstenning** [Pylons/pyramid#1591]
As discussed yesterday on IRC with @mmerickel.

[The ability to alter the virtual root](http://docs.pylonsproject.org/projects/pyramid/en/latest/narr/vhosting.html#virtual-root-support) of a traversal-based pyramid application using the `HTTP_X_VHM_ROOT` WSGI environment variable (and hence the `X-Vhm-Root` HTTP header) could open up some pyramid applications (those using traversal) to cache poisoning attacks.

Specifically, I don't think it's reasonable to assume that operators deploying pyramid applications will know to filter this header in requests from untrusted clients. It's certainly not as widely known as `X-Forwarded-For` and friends. As such, anyone running a reverse-proxy cache in front of a pyramid application could easily find themselves with a poisoned cache if someone makes a request such as

    curl -H 'X-Vhm-Root: /foo/bar' http://example.com/

Perhaps support for this header should be off by default?

**Comment by mmerickel** [comment]
I think it would not be very inconvenient for anyone currently depending on that feature to opt-in. Ideally I think we would split this into a tween that people could include in their apps.

**Comment by tilgovi** [comment]
Tween seems easy. It would just need to modify `PATH_INFO`, right? That's what the Paste virtual host middleware does, I think.

**Comment by mmerickel** [comment]
Currently there's a few spots in pyramid that use the header. Instead of changing all of those, I think I'd rather just add a tween (or something else) to pyramid's pipeline that filters the header based on a setting. For example:

```
pyramid.enable_vhm = true
```

We could also consider something like `pyramid.trusted_proxy` that accepts an ip address, but I think I'd rather just have it on or off. Can anyone else weigh in on this?

**Comment by mcdonc** [comment]
I think we should tackle this first as a documentation task which describes how to build a tween (or newrequest subscriber) that filters out X-Vhm-Root, X-Forwarded-For if REMOTE_ADDR is not trusted.  Alternately, document an Apache and/or nginx config that sets-and/or-deletes these headers, cleaning the environment before proxying (or just schlepping in the case of mod_wsgi).

Once that documentation is done, we can decide whether to add software to Pyramid to make the task easier.

Note that right now Webob also treats this as a documentation task: request.client_address returns the value of HTTP_X_FORWARDED_FOR unconditionally without any notion of having a trusted proxy list.  I'd slightly rather extend Webob to have the concept of a list of trusted proxies, and a list of HTTP vars that should be cleaned from the environ than put this concept directly into Pyramid.

**Comment by mmerickel** [comment]
Yeah so the "pyramid" solution for this is that we need a hardening chapter in the docs that makes it clear what headers pyramid uses.

The rest of the solution is having your proxy or other server remove these untrusted headers.

**Comment by bertjwregeer** [comment]
I'd be okay with adding middleware to WebOb (not enabled by default, mind you) that does sanitization of the environment, there have been a couple of such requests already, with one PR that already contains some of it.

**Comment by mmerickel** [comment]
This should be documented more clearly via #1845.
